packages/flutter_tools/lib/src/localizations/message_parser.dart

```diff
@@ -0,0 +1,510 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// Symbol Types
+import 'gen_l10n_types.dart';
+
+enum ST {
+  // Terminal Types
+  openBrace,
+  closeBrace,
+  comma,
+  equalSign,
+  other,
+  plural,
+  select,
+  string,
+  number,
+  identifier,
+  empty,
+  // Nonterminal Types
+  message,
+
+  placeholderExpr,
+
+  pluralExpr,
+  pluralParts,
+  pluralPart,
+
+  selectExpr,
+  selectParts,
+  selectPart,
+}
+
+// The grammar of the syntax.
+Map<ST, List<List<ST>>> grammar = <ST, List<List<ST>>>{
+  ST.message: <List<ST>>[
+    <ST>[ST.string, ST.message],
+    <ST>[ST.placeholderExpr, ST.message],
+    <ST>[ST.pluralExpr, ST.message],
+    <ST>[ST.selectExpr, ST.message],
+    <ST>[ST.empty],
+  ],
+  ST.placeholderExpr: <List<ST>>[
+    <ST>[ST.openBrace, ST.identifier, ST.closeBrace],
+  ],
+  ST.pluralExpr: <List<ST>>[
+    <ST>[ST.openBrace, ST.identifier, ST.comma, ST.plural, ST.comma, ST.pluralParts, ST.closeBrace],
+  ],
+  ST.pluralParts: <List<ST>>[
+    <ST>[ST.pluralPart, ST.pluralParts],
+    <ST>[ST.empty],
+  ],
+  ST.pluralPart: <List<ST>>[
+    <ST>[ST.identifier, ST.openBrace, ST.message, ST.closeBrace],
+    <ST>[ST.equalSign, ST.number, ST.openBrace, ST.message, ST.closeBrace],
+    <ST>[ST.other, ST.openBrace, ST.message, ST.closeBrace],
+  ],
+  ST.selectExpr: <List<ST>>[
+    <ST>[ST.openBrace, ST.identifier, ST.comma, ST.select, ST.comma, ST.selectParts, ST.closeBrace],
+    <ST>[ST.other, ST.openBrace, ST.message, ST.closeBrace],
+  ],
+  ST.selectParts: <List<ST>>[
+    <ST>[ST.selectPart, ST.selectParts],
+    <ST>[ST.empty],
+  ],
+  ST.selectPart: <List<ST>>[
+    <ST>[ST.identifier, ST.openBrace, ST.message, ST.closeBrace],
+    <ST>[ST.other, ST.openBrace, ST.message, ST.closeBrace],
+  ],
+};
+
+class Node {
+  Node(this.type, this.positionInMessage, { this.expectedSymbolCount = 0, this.value, List<Node>? children }): children = children ?? <Node>[];
+
+  // Token constructors.
+  Node.openBrace(this.positionInMessage): type = ST.openBrace, value = '{';
+  Node.closeBrace(this.positionInMessage): type = ST.closeBrace, value = '}';
+  Node.brace(this.positionInMessage, String this.value) {
+    if (value == '{') {
+      type = ST.openBrace;
+    } else if (value == '}') {
+      type = ST.closeBrace;
+    } else {
+      // We should never arrive here.
+      throw L10nException('Provided value is not a brace.');
```

Comment:
Could include the value in the message. Less confusing in case this ever comes up.

Reply:
Fixed!